File: StaticSharp/CrossplatformLibrary/Num/Num.cs
using Javascriptifier;
using StaticSharp.Animations;

namespace StaticSharp {


    namespace Scripts {
        public class NumAttribute : ScriptReferenceAttribute {
            public NumAttribute() : base(GetScriptFilePath()) { }
        }
    }




    namespace Js {

        public static class Constants {
            public static readonly string Undefined = "undefined";
        }


        //[JavascriptClass("")]
        public static class Num {
            [JavascriptOnlyMember]
            public static double First(params double[] value) => throw new JavascriptOnlyException();

            [JavascriptOnlyMember]
            public static double Sum(params double[] value) => throw new JavascriptOnlyException();

            [JavascriptOnlyMember]
            public static double Min(params double[] value) => throw new JavascriptOnlyException();

            [JavascriptOnlyMember]
            public static double Max(params double[] value) => throw new JavascriptOnlyException();

            [JavascriptOnlyMember]
            public static double Clamp(double value, double min, double max ) => throw new JavascriptOnlyException();

            public static double PiecewiseLinearInterpolation(double x, params (double x, double y)[] keyframes) {
                if (keyframes.Length == 0) {
                    throw new ArgumentException("Keyframes list must not be empty.");
                }

                if (keyframes.Length == 1) {
                    return keyframes[0].y;
                }

                // Sort keyframes by x.
                Array.Sort(keyframes, (a, b) => a.x.CompareTo(b.x));

                if (x < keyframes[0].x) {
                    return keyframes[0].y;
                }

                if (x >= keyframes[keyframes.Length - 1].x) {
                    return keyframes[keyframes.Length - 1].y;
                }

                for (int i = 1; i < keyframes.Length; i++) {
                    if (x < keyframes[i].x) {
                        double deltaX = keyframes[i].x - keyframes[i - 1].x;
                        double alpha = (x - keyframes[i-1].x) / deltaX;
                        double deltaY = keyframes[i].y - keyframes[i - 1].y;
                        return keyframes[i-1].y + deltaY * alpha;
                    }
                }

                // This should never happen if the keyframes are sorted correctly.
                throw new Exception("Keyframes are not sorted correctly.");
            }

        }

    }


}

File: StaticSharp/CrossplatformLibrary/Num/Num.js

var Num = {}

Num.ValidNumber = function (value) {
    if (typeof (value) !== "number") return false
    return !isNaN(value)
}

Num.IsNaNOrNull = function (value) {
    if (value === null) return true
    return isNaN(value)
}

Num.Max = function () {
    let result = undefined
    for (let i of arguments) {
        if (Num.IsNaNOrNull(i)) continue;
        if (result === undefined) {
            result = i
        } else {
            result = Math.max(result, i)
        }
    }
    return result
}

Num.Min = function () {
    let result = undefined
    for (let i of arguments) {
        if (Num.IsNaNOrNull(i)) continue;
        if (result === undefined) {
            result = i
        } else {
            result = Math.min(result, i)
        }
    }
    return result
}

Num.Clamp = function (value, min, max) {
    return Num.Max(min, Num.Min(max, value))
}

Num.Sum = function () {
    let resultValid = false
    let result = 0
    for (let i of arguments) {
        if (Num.IsNaNOrNull(i)) continue;
        resultValid = true
        result += i
    }
    return resultValid ? result : undefined
}


Num.First = function () {
    for (let i of arguments) {
        if (i === undefined) continue;
        if (isNaN(i)) continue;
        if (i === null) continue;
        return i
    }
    return undefined
}

Num.PiecewiseLinearInterpolation = function (x, ...keyframes) {

    if (x == undefined)
        return undefined

    if (keyframes.length === 0) {
        throw new Error("Keyframes list must not be empty.");
    }

    if (keyframes.length === 1) {
        return keyframes[0][1];
    }

    // Sort keyframes by x.
    keyframes.sort((a, b) => a[0] - b[0]);
    
    if (x < keyframes[0][0]) {
        return keyframes[0][1];
    }

    if (x >= keyframes[keyframes.length - 1][0]) {
        return keyframes[keyframes.length - 1][1];
    }

    for (let i = 1; i < keyframes.length; i++) {
        if (x < keyframes[i][0]) {
            const deltaX = keyframes[i][0] - keyframes[i - 1][0];
            const alpha = (x - keyframes[i - 1][0]) / deltaX;
            const deltaY = keyframes[i][1] - keyframes[i - 1][1];
            return keyframes[i - 1][1] + deltaY * alpha;
        }
    }

    // This should never happen if the keyframes are sorted correctly.
    throw new Error("Keyframes are not sorted correctly.");
}